### PR TITLE
CXX-690: support bypass document validation in agg and bulk ops

### PR DESCRIPTION
**Pull request by Machyne**
NOTE: this is not implemented in find and modify (`find_one_and_update` and `find_one_and_replace`). We are waiting on the C Driver for this.

### CODE REVIEW

File: src/mongocxx/options/aggregate.hpp [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2014 MongoDB Inc.
+// Copyright 2015 MongoDB Inc.
```
Comment (samantharitter) [inline src/mongocxx/options/aggregate.hpp]: Leave this as 2014, I think

File: src/mongocxx/options/aggregate.hpp [hunk 2]
```diff
@@ -21,6 +21,7 @@
 #include <bsoncxx/document/view.hpp>
 #include <bsoncxx/stdx/optional.hpp>
 #include <mongocxx/read_preference.hpp>
```
Comment (samantharitter) [inline src/mongocxx/options/aggregate.hpp]: In the stdx PR, you've changed most things to have a space between the bsoncxx includes and the mongocxx includes.  Either style is fine with me, but let's be consistent.

File: src/mongocxx/bulk_write.hpp [hunk 3]
```diff
@@ -101,6 +101,14 @@ class MONGOCXX_API bulk_write {
     MONGOCXX_DEPRECATED void* implementation() const;
 
     ///
+    /// Whether or not to bypass document validation for this operation.
+    ///
+    /// @param bypass_document_validation
+    ///   TWhether or not to bypass document validation.
```
Comment (markbenvenuto) [inline src/mongocxx/bulk_write.hpp]: TWhether - spelling

Comment (Machyne) [inline src/mongocxx/bulk_write.hpp]: Not sure what that was... thanks x)

File: src/mongocxx/options/aggregate.hpp (hunk 2 above)
Comment (Machyne) [inline src/mongocxx/options/aggregate.hpp]: Nice catch! Space added.

**Comment by Machyne** [comment]
Waiting on [CDRIVER-971](https://jira.mongodb.org/browse/CDRIVER-971) for tests to pass.

File: .travis.yml [hunk 4]
```diff
@@ -58,7 +58,8 @@ install:
     - sudo apt-get install cmake
 
     # Install Mongo C Driver
-    - pushd mongo-c-driver; ./autogen.sh --enable-tests=no --enable-examples=no; make; sudo make install; popd;
+    - mkdir -p /var/tmp/cdriver
+    - pushd mongo-c-driver; ./autogen.sh --enable-tests=no --enable-examples=no --prefix=/var/tmp/cdriver; make; sudo make install; popd;
```
Comment (samantharitter) [inline .travis.yml]: Is this rebased against my recent travis changes?  Wondering because `--with-libbson=bundled` isn't here.

Comment (Machyne) [inline .travis.yml]: Sorry, just rebased now.

File: src/mongocxx/collection.cpp [hunk 5]
```diff
@@ -326,6 +340,7 @@ bsoncxx::stdx::optional<bsoncxx::document::value> collection::find_one_and_repla
     options::return_document rd =
         options.return_document().value_or(options::return_document::k_before);
 
+    // TODO handle bypass document validation when C driver offers this for FAM
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: Can you link the associated C-driver ticket in the comment please?

File: src/mongocxx/options/aggregate.cpp [hunk 6]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2014 MongoDB Inc.
+// Copyright 2015 MongoDB Inc.
```
Comment (samantharitter) [inline src/mongocxx/options/aggregate.cpp]: will most of these bsoncxx::stdx -> stdx changes go away once rebased?

File: src/mongocxx/options/find_one_and_replace.hpp [hunk 7]
```diff
@@ -137,12 +137,33 @@ class MONGOCXX_API find_one_and_replace {
     ///
     const bsoncxx::stdx::optional<bool>& upsert() const;
 
+    ///
+    /// Sets the bypass_document_validation option.
+    /// If true, allows the write to opt-out of document level validation.
+    ///
+    /// @note
+    ///   On servers >= 3.2, unset defaults to false. On servers < 3.2, this
```
Comment (samantharitter) [inline src/mongocxx/options/find_one_and_replace.hpp]: I'd take out the work "unset"

File: src/mongocxx/options/find_one_and_update.hpp [hunk 8]
```diff
@@ -136,12 +136,33 @@ class MONGOCXX_API find_one_and_update {
     ///
     const bsoncxx::stdx::optional<bool>& upsert() const;
 
+    ///
+    /// Sets the bypass_document_validation option.
+    /// If true, allows the write to opt-out of document level validation.
+    ///
+    /// @note
+    ///   On servers >= 3.2, unset defaults to false. On servers < 3.2, this
```
Comment (samantharitter) [inline src/mongocxx/options/find_one_and_update.hpp]: same notes from documentation above

File: src/mongocxx/options/insert.hpp [hunk 9]
```diff
@@ -32,6 +32,26 @@ class MONGOCXX_API insert {
 
    public:
     ///
+    /// Sets the bypass_document_validation option.
+    /// If true, allows the write to opt-out of document level validation.
+    ///
+    /// @note
+    ///   On servers >= 3.2, unset defaults to false. On servers < 3.2, this
```
Comment (samantharitter) [inline src/mongocxx/options/insert.hpp]: ditto here

File: src/mongocxx/test/collection_mocked.cpp [hunk 10]
```diff
@@ -101,6 +102,11 @@ TEST_CASE("Collection", "[collection]") {
                             expected_batch_size);
                 }
 
+                if (opts.bypass_document_validation())
+                    REQUIRE(o["bypassDocumentValidation"].get_bool().value == expected_bypass_document_validation);
+                else
+                    REQUIRE(o.find("bypassDocumentValidation") == o.end());
```
Comment (samantharitter) [inline src/mongocxx/test/collection_mocked.cpp]: or just `REQUIRE(!o["bypassDocumentValidation"])`

File: src/mongocxx/collection.cpp (hunk 5 above)
Comment (Machyne) [inline src/mongocxx/collection.cpp]: Linked!

File: src/mongocxx/test/collection_mocked.cpp [hunk 11]
```diff
@@ -377,6 +396,12 @@ TEST_CASE("Collection", "[collection]") {
                 options.upsert(upsert_option);
             }
 
+
+            SECTION("Upsert false") {
```
Comment (samantharitter) [inline src/mongocxx/test/collection_mocked.cpp]: Is this change related?

File: src/mongocxx/collection.cpp [hunk 12]
```diff
@@ -316,6 +330,8 @@ stdx::optional<bsoncxx::document::value> collection::find_one_and_replace(
...
+    // @see https://jira.mongodb.org/browse/CDRIVER-945
```
Comment (samantharitter) [inline src/mongocxx/collection.cpp]: The '@thing' syntax is really just for our official header comments that get parsed into documentation.  For this it's fine to just do `// TODO handle bypass document validation when CDRIVER-945 is done` or something

**Comment by samantharitter** [comment]
some small comments, otherwise LGTM

File: src/mongocxx/test/collection_mocked.cpp [hunk 13]
```diff
@@ -106,6 +107,12 @@ TEST_CASE("Collection", "[collection]") {
                             expected_batch_size);
                 }
 
+                if (opts.bypass_document_validation())
+                    REQUIRE(o["bypassDocumentValidation"].get_bool().value ==
+                            expected_bypass_document_validation);
+                else
+                    REQUIRE(o.find("bypassDocumentValidation") == o.end());
```
Comment (Machyne) [inline src/mongocxx/test/collection_mocked.cpp]: @samantharitter 's comment went away, not sure why:
> or just `REQUIRE(!o["bypassDocumentValidation"])`

Should I change it to `!o["key"]` for all the prior tests?

Comment (samantharitter) [inline src/mongocxx/test/collection_mocked.cpp]: eh, not a huge deal

File: src/mongocxx/bulk_write.hpp [hunk 14]
```diff
@@ -101,6 +101,14 @@ class MONGOCXX_API bulk_write {
     MONGOCXX_DEPRECATED void* implementation() const;
 
     ///
+    /// Whether or not to bypass document validation for this operation.
+    ///
+    /// @param bypass_document_validation
+    ///   Whether or not to bypass document validation.
+    ///
+    void bypass_document_validation(bool bypass_document_validation);
```
Comment (acmorrow) [inline src/mongocxx/bulk_write.hpp]: Normally, I'm pretty against bool parameters to functions, but I'm ok with it on a trivial setter when there are no other parameters. No need to change anything, just pointing out that I view this as a special exception.

**Comment by acmorrow** [comment]
Neat. LGTM
